[ptd] make multithreaded pg wait for readiness before the 1st collective

xunnanxu opened the pull request:
Summary:
This used to be not a problem because in c10d collective init, a store based barrier would be applied.

This recently got changed in https://github.com/pytorch/pytorch/pull/103033
where the barrier is not by default applied.

For normal PGs like gloo/nccl, this is not a problem as the rendezvous process is implicitly a barrier anyway.

But for threaded pg, without the store based barrier this would lead to race condition as the local pg does not wait for world to be ready before starting collectives.

This fixes the issue by just doing a store based barrier for each pg created.
The CV attempt wouldn't work since that would still rely on class level variables which would break in the device mesh case. See inline comment for details.

Differential Revision: D48220125

wanchaol commented on torch/testing/_internal/distributed/multi_threaded_pg.py:
```diff
@@ -232,6 +232,7 @@ class ProcessLocalGroup(dist.ProcessGroup):
     _pg_list = []
     _count = 0
     _ready = False
```
can we remove the `_ready` flag too? the whole point of this flag can be substitute by the ready_cond cv.

XilunWu replied:
Yeah we had `_ready` to serve the same purpose but somehow the usage was removed....

xunnanxu replied:
I attempted to remove this flag. It turned out to expose some issues in readiness check.
@wanchaol @XilunWu please check my latest change in this PR as I removed the class vars related to readiness check and simply used store based barrier instead (see inline comment for rationale).

The reason why store based barrier instead of regular barrier is because latter would have chicken-n-egg problem otherwise (cannot execute collective before pg is ready) as the current threaded pg doesn't have its own rendezvous like gloo or nccl.

wanchaol reviewed (approved):
lgtm, one suggestions inlined

H-Huang reviewed (approved):
Thanks for fix. curious, what is the case in which this causes issues? Is there a way to test this?

XilunWu reviewed (approved):
Thx for the fix and it's good to merge once `_ready` is removed.

XilunWu commented:
> Thanks for fix. curious, what is the case in which this causes issues? Is there a way to test this?

We have use of multi-threaded PG in DTensor's test but it wasn't turned off by Github CI. I believe a local run of that should justify the correctness of the fix.

xunnanxu committed:
[ptd] make multithreaded pg wait for readiness before the 1st collective (#106954)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/106954

This used to be not a problem because in c10d collective init, a store based barrier would be applied.

This recently got changed in https://github.com/pytorch/pytorch/pull/103033
where the barrier is not by default applied.

For normal PGs like gloo/nccl, this is not a problem as the rendezvous process is implicitly a barrier anyway.

But for threaded pg, without the store based barrier this would lead to race condition as the local pg does not wait for world to be ready before starting collectives.

This fixes the issue by just doing a store based barrier for each pg created.
The CV attempt wouldn't work since that would still rely on class level variables which would break in the device mesh case. See inline comment for details.

Reviewed By: H-Huang, XilunWu, wanchaol

Differential Revision: D48220125

fbshipit-source-id: 75eb56a028df321331a75c5ffe5b8f7518af356a

xunnanxu commented:
> Thanks for fix. curious, what is the case in which this causes issues? Is there a way to test this?

@H-Huang @XilunWu Basically before the fix, the tests may start executing collectives before all threads are ready (and hence triggering the error. The GH test may not have enough pressure to trigger this but we observe tests would fail under stress (see internal test plan for [D48220125](https://www.internalfb.com/diff/D48220125))

wanchaol reviewed (approved):
the store based barrier change looks good to me, actually I think it's much cleaner to the previous implementation, thanks for resolving this issue!

xunnanxu commented:
@pytorchbot merge

xunnanxu commented:
@pytorchmergebot

xunnanxu commented:
@pytorchbot merge
